ccmbioinfo/lfs
# CARDS-1492: Slack notifications: # of upcoming appts, # of initial emails sent, # of reminder emails sent, # of surveys completed

The PR implements CARDS-1492 by providing various performance counters in the JCR under `/PerformanceStatistics/` and a periodic task which sends updates to a Slack channel with the values of the performance counters.

To test:

Testing setup
------------------

1. Build the _CARDS-1492_ branch (`mvn clean install`)
2. If you are using the provided development _mock Slack_ server, open a new terminal and run `cd Utilities/Development`, then `nodejs mock_slack.js`. Otherwise, if you are using the private _cards-alerts-dev_ channel, please message me and I will add you to this channel and provide you with the private URL to POST to for sending messages to that channel.
3. Start CARDS in _Cards4Proms_ mode with the command `SLACK_PERFORMANCE_URL=http://localhost:8000/ PATIENT_NOTIFICATION_FROM_ADDRESS='datapro.uhn.ca' PATIENT_NOTIFICATION_FROM_NAME='UHN DATAPRO' NIGHTLY_NOTIFICATIONS_SCHEDULE='0 * * * * ? *' SLING_COMMONS_CRYPTO_PASSWORD=password CARDS_HOST_AND_PORT='localhost:8080' NIGHTLY_SLACK_NOTIFICATIONS_SCHEDULE='0 * * * * ? *' ./start_cards.sh --project cards4proms --dev`
4. Every minute, you should see a new Slack message being posted with the _Today_ and _Total_ values for the performance statistics of _Initial Emails Sent_, _Reminder Emails Sent_, _Total Surveys Submitted_, _Appointment Surveys Submitted_, and _Imported Appointments_. Please note that given, for testing purposes, we have set this task to run every minute instead of once a day, a day in this testing scenario is considered to be one minute long, therefore the _Today_ values for a statistic are reset every minute.

Testing _Initial Emails Sent_ Performance Statistic
---------------------------------------------------------------------

1. In a new terminal, start the mock Torch server for an appointment 3 days from now by running `cd Utilities/Development`, then `nodejs mock_torch.js --appointment-time-hours-from-now=72`.
2. Visit `http://localhost:8080/system/console/configMgr`.
3. Look for the `PROMs import` configuration, click on it to add a new config, set the _Name_ to `mock`, _Endpoint URL_ to `http://localhost:8011/`, and _Clinic_ to `6012-HC-Congenital Cardiac`. Save the configuration.
4. Point your browser to `http://localhost:8080/Subjects.importTorch.json?config=mock`.
5. Visit `http://localhost:8080` once again. You should see an imported patient and their scheduled visit.
6. On Slack (or the mock Slack server), a message should be posted for `Initial Emails Sent -- Today: 1, Total: 1`, then one minute later, `Initial Emails Sent -- Today: 1, Total: 2`, then one minute later, `Initial Emails Sent -- Today: 1, Total: 3`, ...etc.

Testing _Reminder Emails Sent_ Performance Statistic
--------------------------------------------------------------------------

1. Stop the `mock_torch.js` server and restart it with the parameter `--appointment-time-hours-from-now=24` (instead of `--appointment-time-hours-from-now=72`).
2. Point your browser to `http://localhost:8080/Subjects.importTorch.json?config=mock`. This import should cause the initially imported appointment to be moved from 3 days from now to 1 day from now.
3. On Slack (or the mock Slack server), a message should be posted for `Reminder Emails Sent -- Today: 1, Total: 1`, then one minute later, `Reminder Emails Sent -- Today: 1, Total: 2`, then one minute later, `Reminder Emails Sent -- Today: 1, Total: 3`, ...etc.

Testing _Total Surveys Submitted_ Performance Statistic
------------------------------------------------------------------------------

1. Visit `http://localhost:8080/system/console/configMgr`.
2. Look for the `Submission Counter` configuration, click on it to add a new config, and save the default values.
3. Open the imported _Visit information_ Form and set the _completed_ and _submitted_ values to _Yes_. Save the Form.
4. On Slack (or the mock Slack server), a message should be posted for `Total Surveys Submitted: Today: 5, Total: 5` then `Total Surveys Submitted: Today: 0, Total: 5` then `Total Surveys Submitted: Today: 0, Total: 5`, ...etc.

Testing _Appointment Surveys Submitted_ Performance Statistic
-----------------------------------------------------------------------------------------

1. On Slack (or the mock Slack server), a message should be posted for `Appointment Surveys Submitted: Today: 1, Total: 1` then `Appointment Surveys Submitted: Today: 0, Total: 1` then `Appointment Surveys Submitted: Today: 0, Total: 1`, ...etc.

Testing _Imported Appointments_ Performance Statistic
-----------------------------------------------------------------------------

1. Given that we already have imported two appointments, on Slack (or the mock Slack server), further up on the page, a message should be posted for `Imported Appointments: Today: 1, Total: 1` then `Imported Appointments: Today: 0, Total: 1` then `Imported Appointments: Today: 0, Total: 1`, ..., `Imported Appointments: Today: 1, Total: 2` then `Imported Appointments: Today: 0, Total: 2` then `Imported Appointments: Today: 0, Total: 2`, ...etc.


Please note that the _Imported Appointments_ performance counter increments each time the Torch import task runs and obtains upcoming visits and not when new visits are created or existing visits are modified. Should this behavior be changed? @marta-
